### ISSUE TITLE
Unnecessary restrictions on ranges of values

### ISSUE THREAD

**Issue by isaac-dunn** [cambridge-alpha-team/unvisual-frontend#44]
Sorry to keep going on about this (cf. discussion in #17), but I really do think a lot of our ranges of values are unnecessarily restrictive. It was previously justified because we thought the cubelet outputs would be mapped to these moderate ranges, but as I understand it, we're now letting the user specify the range to which the cubelet outputs will be mapped.

Therefore, I think that restrictions should be made as lenient as possible. I suggest the following ranges:
- Sleep: 0.125 - 1000
- Play: 25 (inaudible on even excellent equipment) - 130 (sounds so bad it could never be musical)
- Tempo: 5 - 1000
- Amp: 0.125 - 10
- Release: 0.125 - 100 (aside: are we ever going to allow attack, sustain, etc?)

There are probably others that I've missed here.

**Comment by RusticFlare** [comment]
Amp should never be heigher than 1 (see the sonic pi documentation on how it works). I also think that for this project restrictions are good because if no music is produced then the child will be confused and using the cubelets will produce awful (or inaudable) sounds when they go to extreme values.

**Comment by LauraneS** [comment]
I agree with that. What is the point of letting people puting values in that don't produce anything?

**Comment by isaac-dunn** [comment]
> However, you should note that when the overall amplitude of all the sounds gets too high, Sonic Pi uses what’s called a compressor to squash them all to make sure things aren’t too loud for your ears. This can often make the sound muddy and strange. So try to use low amplitudes, i.e. in the range 0 to 0.5 to avoid compression. 

As far as I can see, the Sonic Pi documentation only _recommends_ that you don't use amplitudes greater than one, because you might not like the sound, which in my opinion is not enough reason to forbid users from ever setting amplitudes greater than one. I thought we were trying to give the user as much power as possible, which to me means not telling them what sounds they will and won't like.

My point about the cubelets was that the cubelet output value will not map onto the entire possible range of the parameter being set, but instead onto some smaller range that the user specifies. Or have I misunderstood, and this is not the case?

The reason I have included inaudible/awful-sounding values is because as far as I can see that is the only way to include all possible values that are **not** inaudible/awful-sounding. Anyone going to these extremes would surely only be doing it for the sake of it, and so I would argue that they would be fairly likely to realise what's wrong when it's inaudible/screechy.

What ranges would you propose instead, if you disagree with mine?

**Comment by tjvr** [comment]
> we're now letting the user specify the range to which the cubelet outputs will be mapped.

That sounds like a bad idea...

**Comment by RusticFlare** [comment]
The documentation I meant:

> The amplitude of the sound. Typically a value between 0 and 1. Higher amplitudes may be used, but won't make the sound louder, it will just reduce the quality of all the sounds currently being played (due to compression.)

**Comment by isaac-dunn** [comment]
> That sounds like a bad idea...

Care to elaborate?

**Comment by isaac-dunn** [comment]
@RusticFlare Fair enough, I agree, amplitude should be greater than 0 and at most 1. Any thoughts on the rest of my comment?

**Comment by RusticFlare** [comment]
I think sleep and release need to be longer than they currently are, but 1000 is too long.
Play should be limited to what you can hear though a laptop.
I'm not sure about tempo...

**Comment by isaac-dunn** [comment]
Why is 1000 too long?

Why should play be limited to what you can hear through a laptop? People will want to use our system with headphones or good quality speakers (e.g. in a live coding performance), so it doesn't make much sense to me to restrict the values they are allowed to use just because _if_ it was on a laptop, it wouldn't be audible.

**Comment by RusticFlare** [comment]
The whole point of the limits is to make sure that the code works and sounds sensible. Otherwise we may as well have no limits.

**Comment by RusticFlare** [comment]
The reason for limiting it to a laptops audio range is you may make a piece that you can hear through earphones, but it fails when you try to share it with your laptop.

**Comment by tjvr** [comment]
> Care to elaborate?

Sure, I'd be happy to. I'm just a little confused, because your original comment seems to be referencing a discussion I can't find:

> as I understand it, we're now letting the user specify the range to which the cubelet outputs will be mapped.

Allowing the user to set the range for the cubelets would add complexity to the implementation and to the user interface. The user has to learn and keep in mind two different sets of ranges; and has to learn how to change them. I'd expect users would simply use the default ranges. Giving the user feedback on how the cubelets affect the value is another challenge.

I'd also emphasise the value of sensible defaults, and reasonable limits. I don't see any justification in making the min/max values overly large. I think it's reasonable to assume reasonable-quality audio equipment; but at the same time, having clear feedback is important. As I [wrote before](https://github.com/cambridge-alpha-team/unvisual-frontend/issues/17#issuecomment-74437317):

> This matters because the user will struggle to work out what's wrong (have they used the "play" command correctly?), which may impede their learning or make them frustrated.

**Comment by LauraneS** [comment]
We need to reach an agreement here, and this discussion has to do with #51 as well.

**Comment by isaac-dunn** [comment]
Well, everyone seems to disagree with me, so if we go with the majority then we shouldn't change it.

I do still strongly feel that the restrictions on sleep, play, tempo and release should be weakened, even if not as much as I originally suggested.
